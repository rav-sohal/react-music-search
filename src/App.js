import React, {Component} from 'react';
import axios from 'axios';
import Header from './components/Header';
import Search from './components/Search';
import Results from './components/Results';

const terms = [
  'drake',
  'balvin',
  'black eyes peas',
  'madonna',
  'weekend',
  'khalid',
];

class App extends Component {

  state = {
    albums: [],
    loading: true,
    title: true, 
    errorMessage: ''
  };

  componentDidMount () {
    this.initValue();
  }

  // componentDidMount () {
  //   fetch ('https://itunes.apple.com/search?term=drake&enitity=album&limit=10') // fetch data
  //     .then (res => res.json ()) // parse to json
  //     .then (responseData => {
  //       // take json data and insert into state using callback
  //       this.setState ({albums: responseData.results});
  //     })
  //     .catch (error => {
  //       // catch errors
  //       console.log ('Error fetching and parsing data', error);
  //     });
  // }

  randomInitValue (element = []) {
    const randomElement = element[Math.floor (Math.random () * element.length)];
    return randomElement;
  }

  initValue = (searchQuery = this.randomInitValue (terms)) => {
    axios
    .get (
      `https://itunes.apple.com/search?term=${searchQuery}&entity=album&limit=10`
    ) // fetch data from iTunes
    .then (response => {
      // axios returns data in json, so just need to upate state
      this.setState (prevState => ({
        // update the album array state to the response
        albums:  response.data.results,
        loading: prevState.false,
      }));
    })
    .catch (error => {
      // catch errors
      if (error) {
      console.log ('Error fetching and parsing data', error);
      this.setState (prevState => ({
        loading: prevState.false,
        errorMessage: 'This is an error' + error
      }));
    } 
    });
    
  }

  performSearch = (searchQuery) => {
    axios
      .get (
        `https://itunes.apple.com/search?term=${searchQuery}&entity=album&limit=10`
      ) // fetch data from iTunes
      .then (response => {
        // axios returns data in json, so just need to upate state
        this.setState ({
          // update the album array state to the response
          albums: response.data.results,
          loading: false,
        });
      })
      .catch (error => {
        // catch errors
        console.log ('Error fetching and parsing data', error)       
      });
      this.setState (prevState => ({
        title: prevState.false,
        errorMessage: prevState.false
      }));
  };

  render () {
    return (
      <div>
        <Header />
        <Search onSearch={this.performSearch} />
        {this.state.title ? <div className="container title"><h4>Todays random selection</h4><hr /></div> : <div className="container"><h4>Results</h4><hr /></div>}
        {this.state.errorMessage ? <div className="container"><h4>Error fetching data, please refresh browser {this.state.errorMessage}</h4></div> : <div className="container"></div>}
        {
        this.state.loading
          ? <div className="container text-center mt-3">
              <div className="spinner-border text-secondary">
                <span className="sr-only text-center">Loading...</span>
              </div>
            </div>
          : <Results data={this.state.albums} />}
      </div>
    );
  }
}

export default App;
