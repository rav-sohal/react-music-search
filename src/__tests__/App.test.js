import {render, screen} from '@testing-library/react';
import {create} from 'react-test-renderer';
import App from '../App';

describe ('App.js tests', () => {

  test ('renders Search iTunes Music', () => {
    render (<App />);
    const titleElement = screen.getByText (/Search iTunes Music/i);
    expect (titleElement).toBeInTheDocument ();
  });

  test ('Snapshot test of App', () => {
    const testAppSnapshot = create(<App />).toJSON();
    expect (testAppSnapshot).toMatchSnapshot();
  });

});
