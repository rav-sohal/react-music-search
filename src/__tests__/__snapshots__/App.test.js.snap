// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.js tests Snapshot test of App 1`] = `
<div>
  <nav
    className="navbar navbar-expand navbar-light"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="#home"
        style={
          Object {
            "fontSize": "38px",
            "fontWeight": "bolder",
          }
        }
      >
        SYFM
      </a>
      <button
        aria-label="Toggle navigation"
        className="navbar-toggler collapsed"
        onClick={[Function]}
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        aria-expanded={null}
        className="justify-content-end navbar-collapse collapse"
      >
        <span
          className="navbar-text"
        >
          Search your favourite music
        </span>
      </div>
    </div>
  </nav>
  <div
    className="jumbotron"
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <div
      className="container"
    >
      <h1>
        Search iTunes Music
      </h1>
      <p>
        Search your favourite artist, album or song.
      </p>
      <div
        className="sc-bdvvaa gYYXMo"
      >
        <form
          className="form-group"
          onSubmit={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="col-md-9"
            >
              <input
                className="sc-gsDJrp hzTNRY form-control"
                name="search"
                onChange={[Function]}
                placeholder="Search..."
                type="search"
              />
            </div>
            <div
              className="col-md-3"
            >
              <button
                className="btn btn-success btn-lg btn-block sm-mt-3"
                id="submit"
                type="submit"
              >
                Search
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <h4>
      Todays random selection
    </h4>
    <hr />
  </div>
  <div
    className="container text-center mt-3"
  >
    <div
      className="spinner-border text-secondary"
    >
      <span
        className="sr-only text-center"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
